### PR TITLE
Panel loading indicator

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
This PR contains functionality to make any panel, i.e. layout, pane or widget, look like it's
loading and disabled.

This is very important for the user experience. A user that sees an unresponsive page without any
indication of something happening in the background will either perceive that the page is slow or
be in doubt if the page has stopped working.

For me personally this has been one of the main issues with Bokeh/ Panel. A lot of users I know perceives it as slow because of this. and old fashion. And Bokeh/ Panel can in fact be very, very fast and modern if used correctly.

It's also indirectly requested here https://github.com/holoviz/panel/issues/1362

@philippjfr . The functionality and test is there. I would really like some feedback from you on

- Do you want this PR?
- Do you want to add `start_spinning`, `.stop_spinning` methods to all layouts, panes and widgets? If yes which class should I add it to? And what should we call them? An alternative could be `.start_active_indicator`, `stop_active_indicator`.
- Which spinner do you want`? Please provide link. And where should I put it?
- Should the height and width of the spinner be absolute? Or relative to the `div` container? Relative might be nice for large panels. But a page with many panels of different size can also become very "busy" if all the indicators have different sizes.
- Should this functionality really go to Bokeh? FYI @mattpap and @bryevdv 

You can run the app below via `python -m panel serve panel/tests/test_loading_indicator.py --dev`

There is a high resolution video here https://discourse.holoviz.org/t/loading-spinner-for-layouts-panes-and-widgets-how-should-it-look-and-feel/1370

loading_spinners

Feel free to finalize this PR if you are up for it 😄 

------

Todo:

- [ ] Maybe add `start_spinning`, `.stop_spinning` methods to all layouts, panes and widgets
- [x] Document
- [x] Select the best spinner, layout (Size, Absolute/ Relative etc.) and style.
- [x] Update CSS across themes. For example for Dark Themes.

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
As of now the best spinner in my opinion is. But I would like it to be a little bit less gray.

I don't think we should have on with color as then users who try to brand their apps using custom colors have a hard time.

Super.

File: panel/loading_indicator.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+background-position: center;
+background-size: 40px 40px;
+border-width: 1px;
+
+}
+.bk.panel-loading:hover:before {
+    cursor: progress;
+}
+"""
+pn.config.raw_css.append(STYLE)
```
Comment (MarcSkovMadsen) [inline panel/loading_indicator.py:32]: @philippjfr . I would need some guidance on where to put the style? And how to add it to Panel.

**Comment by slamer59** [comment]
I made Bokeh rolling for this since it is the underliying technologies for graphs. So depending on what will be loaded I will choose the official logo, rolling if round, fading if rectangle. 

For the button I would not apply a loading since a button is async action in my mind. 
Small stuff :)

**Comment by philippjfr** [comment]
>Do you want to add start_spinning, .stop_spinning methods to all layouts, panes and widgets? If yes which class should I add it to? And what should we call them? An alternative could be .start_active_indicator, stop_active_indicator.

No, I'd probably add a Boolean parameter called `loading`, which can be toggled on or off.
 
>Which spinner do you want`? Please provide link. And where should I put it?

No strong opinions but I'll look at this, and would suggest we offer a few choices and make it configurable globally.

>Should the height and width of the spinner be absolute? Or relative to the div container? Relative might be nice for large panels. But a page with many panels of different size can also become very "busy" if all the indicators have different sizes.

I'd probably vote to make it relative but also globally configurable.

>Should this functionality really go to Bokeh?

Not sure, but I'm happy to merge this and decide whether or not to upstream it later.

**Comment by MarcSkovMadsen** [comment]
Hi @slamer59 

It is you as a developer who decides which layouts, panes or widgets indicates loading or not. So you decide which.

Regarding the bokeh logo I would opt strongly against this as then there is one more thing to change if you want or need to style to your corporate style. We should choose something general. The bokeh logo looks modern on a dark background but old style on a light in my personal opinion.

**Comment by philippjfr** [comment]
I would also like for panes to accept functions directly which are evaluated when the pane is rendered and while it is evaluating it will automatically set the loading parameter but that's a separate PR.

**Comment by MarcSkovMadsen** [comment]
Been exploring https://loading.io for loading spinners. I really like the `bar chart` loading spinner. It's great for loading plots I believe. But maybe not so good for a loading button.

Checkout https://github.com/MarcSkovMadsen/awesome-panel-assets/tree/master/spinners

You can create SVG loading spinners there. So maybe it's possible to chose one or more spinners looks and let the user specify the color.

**GIF**

GIF

**SVG**

SVG

**Comment by MarcSkovMadsen** [comment]
#### Background

I've played a bit around to get an understanding on if and how it would be possible to provide spinner options to users.

The loading indicator is provided via css

```python
.bk.panel-loading:before {
    position: absolute;
    height: 100%;
    width: 100%;
    content: '';
    z-index: 1000;
    background-color: rgb(255,255,255,0.50);
    border-color: lightgray;
    background-image: url('https://raw.githubusercontent.com/holoviz/panel/master/panel/assets/spinner.gif');
    background-repeat: no-repeat;
    background-position: center;
    background-size: 40px 40px;
    border-width: 1px;

}
.bk.panel-loading:hover:before {
    cursor: progress;
}
```

## Learnings

- You can use any type of image in including `svg` in the `background-image: url`.
- But you cannot change the color of the background svg. If you want to enable the user to set the color you need to produce and inline the `svg` string in the url.

## Options

I see the following options

- Develop functionality to enable the user to set the color or even svg image. But then we begin to have a lot of `style` options on the `BasicTemplate`. Maybe they should be refactored into a `style_settings` pararameter?
- Choose one fixed svg spinner for light and one for dark. Document how to change the spinner via css.

## Proposed Options

I would propose fixing a spinner for no BasicTemplate, BasicTemplate Default and BasicTemplate Dark situations.
- Document how to change the spinner via css
- Provide functionality in the BasicTemplate to provide rawcss, css_files, js_files, js_module_files and js_bottom_script locally for that template. This will enable super users to  customize BasicTemplate without creating a new html template.

**Comment by MarcSkovMadsen** [comment]
I've tried to add a `loading` parameter to `Viewable` but it raises an exception when running tests and apps that I don't know how to solve. I thought I needed to add `loading` to `_rename` in one or many places. But I could not find out where.

image

image

**Comment by MarcSkovMadsen** [comment]
Checkout how it works in this short video https://youtu.be/7dJO4j4ENhg

What is left is

- Get CSS into all templates. For the BasicTemplates I know where. But for the *normal* template I don't know where.
- Refactor `loading` property to `loading parameter`. I cannot get this working. See comment above.
- Find out where documentation should go. Maybe some documentation should go to https://panel.holoviz.org/user_guide/Components.html?
- Find out what more Philipp wants me to do before this can be approved.

@philippjfr : In order to finish this I would need your feedback and help to answer the above questions. I don't expect to do more before I get that feedback. Thanks.

File: panel/io/loading.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+STYLE = """
+.bk.pn-loading:before {
+    position: absolute;
+    height: 100%;
+    width: 100%;
+    content: '';
+    z-index: 1000;
+    background-color: rgb(255,255,255,0.50);
+    border-color: lightgray;
+    background-image: url('https://raw.githubusercontent.com/MarcSkovMadsen/awesome-panel-assets/e6cb56375bb1c436975e09739a231fb31e628a63/spinners/default.svg');
```
Comment (philippjfr) [inline panel/io/loading.py:25]: We'll have to inline this or load it from somewhere.

**Comment by philippjfr** [comment]
Add section to `Customization` section and add it to at least a few examples in the gallery.

File: panel/io/resources.py [hunk 3]
```diff
@@ -40,21 +41,51 @@ def conffilter(value):
...
+def loading_css():
+    from ..config import config
+    with open(ASSETS_DIR / f'{config.loading_spinner}_spinner.svg', encoding='utf-8') as f:
+        svg = f.read().replace('\n', '').format(color=config.loading_color)
+    b64 = b64encode(svg.encode('utf-8')).decode('utf-8')
+    return f"""
+    .bk.pn-loading.arcs:before {{
+      background-image: url("data:image/svg+xml;base64,{b64}")
+    }}
+    .bk.pn-loading.arcs:before {{
```
Comment (MarcSkovMadsen) [inline panel/io/resources.py:65]: I had to add this css to get the loading indicator working in the notebook example I just added @philippjfr 

I think it should maybe go to the widgets.css file and an updated version should be included in the dark.css files for it to look nice there.

**Comment by MarcSkovMadsen** [comment]
@philippjfr . I've added a notebook example. I don't know where to put it. A few comments on the direction you would like it to take would be appreciated. Then I can do a second round.

If you would add the css needed or specify where you would like me to put it, then it seems close to ready to me.

File: panel/io/resources.py (hunk 3 above)
Comment (philippjfr) [reply panel/io/resources.py:65]: This definitely should not be needed. I'll remove it again.

Comment (philippjfr) [reply panel/io/resources.py:65]: I probably did something wrong.

Comment (philippjfr) [reply panel/io/resources.py]: Forgot to push `panel/dist/css/loading.css`

**Pull request merged**